sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -781,252 +757,321 @@ public virtual async Task StopProcessingAsync(CancellationToken cancellationToke
+    private static async Task CancelTask(
+            CancellationTokenSource cancellationSource,
+            Task task)
+    {
+        try
+        {
+            if (cancellationSource != null)
             {
-                // Nothing to do here.  These exceptions are expected.
+                cancellationSource.Cancel();
+                await task.ConfigureAwait(false);
             }
         }
+        catch (Exception ex) when (ex is TaskCanceledException)
+        {
+            // Nothing to do here.  These exceptions are expected.
+        }
+    }
 
-        private async Task<ServiceBusReceiver> GetOrCreateSessionReceiver(
-            string sessionId,
-            ServiceBusReceiverOptions receiverOptions,
-            CancellationToken cancellationToken)
+    private async Task<ServiceBusReceiver> GetOrCreateSessionReceiver(
+        string sessionId,
+        ServiceBusReceiverOptions receiverOptions,
+        CancellationToken cancellationToken)
+    {
+        bool releaseSemaphore = false;
+        try
         {
-            ServiceBusReceiver receiver = null;
-            bool releaseSemaphore = false;
-            try
+            if (sessionId == null)
             {
-                // Adding double-checked locking, which checks whether sessionIdsReceiverMap contains sessionId or not.
-                // This is efficient because it avoids creating a receiver for the same session Ids for multiple threads.
-                // Otherwise, we will get session lock lost error whenever multiple threads will try to create a receiver for
-                // the same session id.
-                if (!sessionIdsReceiverMap.ContainsKey(sessionId))
-                {
-                    await receiverCreationLockMap[sessionId].WaitAsync(cancellationToken).ConfigureAwait(false);
-                    releaseSemaphore = true;
+                return await CreateAndInitializeSessionReceiver(
+                    receiverOptions,
+                    cancellationToken,
+                    sessionId)
+                    .ConfigureAwait(false);
+            }
 
-                    if (!sessionIdsReceiverMap.ContainsKey(sessionId))
-                    {
-                        receiver = await CreateAndInitializeSessionReceiver(receiverOptions, cancellationToken, sessionId).ConfigureAwait(false);
-                    }
-                    else
-                    {
-                        receiver = sessionIdsReceiverMap[sessionId];
-                    }
+            // Adding double-checked locking, which checks whether sessionIdsReceiverMap contains sessionId or not.
+            // This is efficient because it avoids creating a receiver for the same session Ids for multiple threads.
+            // Otherwise, we will get session lock lost error whenever multiple threads will try to create a receiver for
+            // the same session id.
+            if (!_sessionIdsReceiverMap.ContainsKey(sessionId))
```

Comment:
This check races with the removal. Receiver can be removed from the map bethween this check returnin true and `_sessionIdsReceiverMap[sessionId]` call

Reply:
Refactored based on your advice.